[libc++] Fix filesystem::remove_all() on FreeBSD

markjdb opened the pull request:
remove_all_impl() opens the target path with O_NOFOLLOW, which fails if
the target is a symbolic link. On FreeBSD, rather than returning ELOOP,
openat() returns EMLINK. This is unlikely to change for compatibility
reasons, see https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=214633 .

Thus, check for EMLINK as well.

markjdb committed:
[libc++] Fix filesystem::remove_all() on FreeBSD

remove_all_impl() opens the target path with O_NOFOLLOW, which fails if
the target is a symbolic link.  On FreeBSD, rather than returning ELOOP,
openat() returns EMLINK.  This is unlikely to change for compatibility
reasons, see https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=214633 .

Thus, check for EMLINK as well.

emaste commented:
The XFAIL was added in 7ec6c629052c30c0772e8fcae346a405af43444e and just captured the current state of the tests in order to stand up the FreeBSD CI instance. It is presumably because of this issue, and assuming the test passes with this change the XFAIL should be removed.

markjdb commented:
> The XFAIL was added in [7ec6c62](https://github.com/llvm/llvm-project/commit/7ec6c629052c30c0772e8fcae346a405af43444e) and just captured the current state of the tests in order to stand up the FreeBSD CI instance. It is presumably because of this issue, and assuming the test passes with this change the XFAIL should be removed.

Do you happen to have a simple recipe to build and run these tests on FreeBSD? I haven't done that before.

emaste commented:
> Do you happen to have a simple recipe to build and run these tests on FreeBSD? I haven't done that before.

I haven't built upstream llvm in a while, but you should just be able to run e.g. `bin/llvm-lit ../libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.remove_all`

It looks a run with this change should have results here in a bit though:
https://buildkite.com/llvm-project/libcxx-ci/builds/33234#018d43a2-514b-40fd-86c4-6e743f60fe21

ldionne reviewed (approved):
This looks reasonable to me but you'll need to remove the XFAIL for the CI to pass. Please also ensure that all the other platforms are green before merging.

markjdb committed:
[libc++][test] Remove XFAIL annotation for the remove_all() test

It passes on FreeBSD now that remove_all_impl() handles FreeBSD's
non-standard error value when opening a symlink with O_NOFOLLOW.

markjdb commented:
> This looks reasonable to me but you'll need to remove the XFAIL for the CI to pass. Please also ensure that all the other platforms are green before merging.

Thank you. I pushed a second commit which does that. I verified that the test now passes on FreeBSD.

DimitryAndric reviewed (approved):
LGTM

markjdb commented:
It looks like one of the test runs was cancelled. Should I try to rerun them?

DimitryAndric commented:
All the buildkite stuff at https://buildkite.com/llvm-project/libcxx-ci/builds/33247#_ says "cancelled after 5h" so I think it's probably some VM or container failing to start. No idea what is behind all this infrastructure. :))

DimitryAndric commented:
Not sure about the CI failure, it just seems like a timeout for a VM to start up. Nothing to do with your commit, as usual.

jwakely commented:
FWIW while handling `EMLINK` in libstdc++ I discovered that NetBSD sets its non-standard `EFTYPE` error for the "openat with O_NOFOLLOW meets a symlink" case.

DragonFly matches FreeBSD, so the current libc++ code should handle that already.

[NetBSD 10.0 `openat`(2)](https://man.freebsd.org/cgi/man.cgi?query=openat&apropos=0&sektion=0&manpath=NetBSD+10.0&arch=default&format=html#ERRORS)

[DragonFly 6.4 `openat`(2)](https://man.freebsd.org/cgi/man.cgi?query=openat&apropos=0&sektion=0&manpath=DragonFly+6.4.0&arch=default&format=html#ERRORS)
